app/services/notification_subscriptions/toggle.rb

```diff
@@ -0,0 +1,94 @@
+module NotificationSubscriptions
+  class Toggle
+    attr_reader :current_user, :permitted_params
+
+    def self.call(...)
+      new(...).call
+    end
+
+    def initialize(current_user, permitted_params)
+      @current_user = current_user
+      @permitted_params = permitted_params
+    end
+
+    def call
+      toggle_subscription
+    end
+
+    def toggle_subscription
+      action = permitted_params[:action]
+
+      if action == "unsubscribe"
+        subscription_id = permitted_params[:subscription_id]
+        notification = NotificationSubscription&.find(subscription_id) unless subscription_id.nil?
+        destroy_notification(notification)
+      else
+        comment_id = permitted_params[:comment_id]
+        article_id = permitted_params[:article_id]
+        comment = Comment.find(comment_id) if comment_id.present?
+        article = Article.find(article_id) if article_id.present?
+
+        create_subscription(comment, article)
+      end
+    end
+
+    private
+
+    def destroy_notification(notification)
+      notification.destroy
+
+      if notification.destroyed?
+        { destroyed: true }
+      else
+        { errors: notification.errors_as_sentence }
+      end
+    end
+
+    def create_subscription(comment, article)
+      comment_article = comment && comment.ancestry.nil? ? comment.commentable : nil
+      notifiable = subscription_notifiable(comment, article, comment_article)
+      notifiable_type = if comment && article.nil? && comment_article.nil?
+                          "Comment"
+                        elsif article.nil?
+                          "Article"
+                        else
+                          "Article"
+                        end
+
+      subscription = NotificationSubscription.new(
+        user: current_user,
+        config: subscription_config(comment, article),
+        notifiable: notifiable,
+        notifiable_type: notifiable_type,
+      )
+
+      if subscription.save
+        { updated: true, notification: subscription.to_json }
+      else
+        { errors: subscription.errors_as_sentence }
+      end
+    end
+
+    def subscription_notifiable(comment, article, comment_article)
+      if comment && article.nil? && comment_article.nil?
+        comment
+      elsif article.nil?
+        comment_article
+      else
+        article
+      end
+    end
+
+    def subscription_config(comment, article)
```

Comment:
The logic here is a little hard to follow. I sort of wonder if it'd be cleaner to pass more explicit instruction from the client-side, but I think we could also work with this. Here's what I think I understand from this code:

* If we got an article_id, subscribe to all of that article's comments
* Or, if we got a comment_id and that comment is not a top comment (has ancestry), subscribe to all the article's comments
* Otherwise (got a comment_id, that comment _is_ a top comment), subscribe to `top_level`

If that's correct, I might suggest we could add a couple of helper methods and stream-line the logic here a bit:

```suggestion
    def subscribing_to_comment?
      article_id.blank? && comment_id.present?
    end

    def subscribing_to_top_comment?
      subscribing_to_comment? && comment.ancestry.present?
    end

    def subscription_config
      subscribing_to_top_comment? ? "top_level_comments" : "all_comments"
```